[kbn/pm] Allow to include/exclude projects in `kbn watch`.

azasypkin opened the pull request:
It should be possible to run `kbn watch` just for a subset of projects, either by including or excluding specific projects (needs to be discussed though). This PR allows developer to specify `-e` (exclude) or `-i` (include) or both:
```bash
$ yarn kbn watch # Will run `kbn:watch` for all known projects
$ yarn kbn watch -i kibana # Will run `kbn:watch` only for `kibana` project
$ yarn kbn watch -e @kbn/pm # Will run `kbn:watch` for every project except for `@kbn/pm`
$ yarn kbn watch -i kibana -i @kbn/datemath # Will run `kbn:watch` only for `@kbn/datemath` and `kibana` projects
```

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1149/)

azasypkin commented:
Hey @kjbekkelund, would you mind taking a look at this WIP PR and let me know what you think?

I'm not 100% sure about multiple inclusion syntax (`-i kibana -i @kbn/datemath`), it's a bit wordy, but supported by default. We can also do `-i kibana,@kbn/datemath` and do splitting on our own if we want.

Also the open question is: do we want to include all dependencies automatically when `-i` is used with the ability to filter out __some__ dependencies with `-e`, e.g. `-i @kbn/es` will automatically run watch for both `@kbn/es` and `@kbn/dev-utils` and `-i @kbn/es -e @kbn/dev-utils` will run watch only for `@kbn/es`. With the current PR  `-i @kbn/es` will run watch __only__ for `@kbn/es` and to run watch for all its dependencies one should include every dependency specifically.

kjbekkelund commented:
I like the "wordy approach". To me it feels a bit clearer.

I think the current way `-i` and `-e` works makes sense, but I think we might want to make them "global" before merging (so they also work for `bootstrap` etc).

Hm, I wonder if `include` is a bad name, as it excludes everything else. Maybe that's why Lerna went with `--scope` and `--ignore` instead of include and exclude. What do you think about the naming?

---

Hm, sidenote: hit a problem when playing with diff. This is my diff:

```diff
diff --git a/packages/kbn-dev-utils/package.json b/packages/kbn-dev-utils/package.json
index 2e2004b3ef..2555564ac4 100644
--- a/packages/kbn-dev-utils/package.json
+++ b/packages/kbn-dev-utils/package.json
@@ -6,7 +6,9 @@
   "private": true,
   "scripts": {
     "build": "babel src --out-dir target",
-    "kbn:bootstrap": "yarn build"
+    "kbn:bootstrap": "yarn build",
+    "watch": "yarn build --watch",
+    "kbn:watch": "yarn watch"
   },
   "dependencies": {
     "chalk": "^2.3.0",
```

Now when running `yarn kbn watch` I keep getting:

```
CliError: Encountered a cycle in the dependency graph. Projects in cycle are:
@kbn/dev-utils
```

But I actually seem to get this on `master` to, so it's not related to this PR, I think.

kjbekkelund commented:
(Haven't looked at code, just had an idea: Oh, I expect the error is because it depends on `@kbn/babel-preset`, but that package is not included in the list of packages. So we might need to re-think when/how we do the topological sort. Maybe do the sort on all first, then extract the relevant packages?)

azasypkin commented:
> I think the current way -i and -e works makes sense, but I think we might want to make them "global" before merging (so they also work for bootstrap etc).

It makes sense, I'll see if I can integrate these options on a global level flawlessly? 

> Hm, I wonder if include is a bad name, as it excludes everything else. Maybe that's why Lerna went with --scope and --ignore instead of include and exclude. What do you think about the naming?

Yeah, it may be confusing, for me it was like:
* `-i` is implicitly enabled by default for __all__ projects
* if you want to exclude some use `-e` (`-i all` is still implicitly enabled)
* if default/implicit behavior doesn't work as needed, then one can explicitly override `-i` only with required project names

But I'm fine with using `-i` for `ignore` and `-s` for `scope` or something similar instead.

> Hm, sidenote: hit a problem when playing with diff. This is my diff:

Thanks, will debug that

kjbekkelund commented:
++ The `-i`, `-e` names work for me

azasypkin commented:
> (Haven't looked at code, just had an idea: Oh, I expect the error is because it depends on @kbn/babel-preset, but that package is not included in the list of packages. So we might need to re-think when/how we do the topological sort. Maybe do the sort on all first, then extract the relevant packages?)

Yeah, that is it. But can we just ignore and not include into batches projects that are not in the `projectsToBatch` [here](https://github.com/elastic/kibana/blob/master/packages/kbn-pm/src/utils/projects.ts#L111)? Or am I missing something?

```ts
const refCounts: { [k: string]: number } = {};
  projects.forEach(pkg =>
    projectGraph.get(pkg.name)!.forEach(dep => {
      // projectsToBatch is a subset of full projectGraph, so no need to count references
      // for the projects that aren't in the target project map.
      if (projectsToBatch.has(dep.name)) {
        if (!refCounts[dep.name]) refCounts[dep.name] = 0;
        refCounts[dep.name]++;
      }
    })
  );
```

Also it looks like `number` from `refCounts: { [k: string]: number }` is actually used as `boolean` - we increment it, but never decrement, just remove the entry entirely. Wondering if we can simplify the code for `topologicallyBatchProjects` to something like this instead. What do you think?
```ts
const projectToBatchNames = new Set(projectsToBatch.keys());

  const batches = [];
  while (projectToBatchNames.size > 0) {
    // Get all projects that have no remaining dependencies within the repo that haven't yet been picked.
    const batch = [];
    for (const projectName of projectToBatchNames) {
      const projectDeps = projectGraph.get(projectName)!;
      const hasNotBatchedDependencies = projectDeps.some(dep => projectToBatchNames.has(dep.name));
      if (!hasNotBatchedDependencies) {
        batch.push(projectsToBatch.get(projectName)!);
      }
    }

    // If we weren't able to find a project with no remaining dependencies,
    // then we've encountered a cycle in the dependency graph.
    const hasCycles = projectToBatchNames.size > 0 && batch.length === 0;
    if (hasCycles) {
      const message =
        'Encountered a cycle in the dependency graph. Projects in cycle are:\n' +
        [...projectToBatchNames].join(', ');

      throw new CliError(message);
    }

    batches.push(batch);

    batch.forEach(pkg => {
      projectToBatchNames.delete(pkg.name);
    });
  }

  return batches;
```

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1677/)

azasypkin commented:
I moved forward and updated PR with changes mentioned in https://github.com/elastic/kibana/pull/17421#issuecomment-380410258 + added some tests (new `topologicallyBatchProjects` tests is failing on master, but passes on this PR).

Let me know what you think @kjbekkelund !

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1688/)

kjbekkelund commented on packages/kbn-pm/src/run.ts:
```diff
@@ -24,6 +24,31 @@ export async function runCommand(command: Command, config: CommandConfig) {
 
     const projects = await getProjects(config.rootPath, projectPaths);
+    const exclude = getExcludeIncludeFilter(config.options.exclude);
+    const include = getExcludeIncludeFilter(config.options.include);
```
What do you think about moving this logic into `getProjects`? E.g. something like this:

```diff
diff --git a/packages/kbn-pm/src/run.ts b/packages/kbn-pm/src/run.ts
index 43cd6e9248..76729c25f7 100644
--- a/packages/kbn-pm/src/run.ts
+++ b/packages/kbn-pm/src/run.ts
@@ -23,22 +23,10 @@ export async function runCommand(command: Command, config: CommandConfig) {
       config.options as ProjectPathOptions
     );
 
-    const projects = await getProjects(config.rootPath, projectPaths);
-    const exclude = getExcludeIncludeFilter(config.options.exclude);
-    const include = getExcludeIncludeFilter(config.options.include);
-
-    // Filter out projects that shouldn't be included if any filters are specified.
-    if (exclude.length > 0 || include.length > 0) {
-      for (const [projectName] of projects) {
-        const excludeProject =
-          exclude.includes(projectName) ||
-          (include.length > 0 && !include.includes(projectName));
-
-        if (excludeProject) {
-          projects.delete(projectName);
-        }
-      }
-    }
+    const projects = await getProjects(config.rootPath, projectPaths, {
+      include: toArray(config.options.include),
+      exclude: toArray(config.options.exclude),
+    });
 
     if (projects.size === 0) {
       console.log(
@@ -82,14 +70,14 @@ export async function runCommand(command: Command, config: CommandConfig) {
   }
 }
 
-function getExcludeIncludeFilter(excludeIncludeRawValue?: string | string[]) {
-  if (excludeIncludeRawValue == null) {
+function toArray(value?: string | string[]) {
+  if (value == null) {
     return [];
   }
 
-  if (typeof excludeIncludeRawValue === 'string') {
-    return [excludeIncludeRawValue];
+  if (typeof value === 'string') {
+    return [value];
   }
 
-  return excludeIncludeRawValue;
+  return value;
 }
diff --git a/packages/kbn-pm/src/utils/projects.ts b/packages/kbn-pm/src/utils/projects.ts
index b465b56531..31b4687ce1 100644
--- a/packages/kbn-pm/src/utils/projects.ts
+++ b/packages/kbn-pm/src/utils/projects.ts
@@ -9,10 +9,12 @@ const glob = promisify(globSync);
 
 export type ProjectMap = Map<string, Project>;
 export type ProjectGraph = Map<string, Project[]>;
+export type ProjectsOptions = { include?: string[]; exclude?: string[] };
 
 export async function getProjects(
   rootPath: string,
-  projectsPathsPatterns: string[]
+  projectsPathsPatterns: string[],
+  { include = [], exclude = [] }: ProjectsOptions = {}
 ) {
   const projects: ProjectMap = new Map();
 
@@ -24,6 +26,12 @@ export async function getProjects(
       const projectDir = path.dirname(projectConfigPath);
       const project = await Project.fromPath(projectDir);
 
+      const excludeProject =
+        exclude.includes(project.name) ||
+        (include.length > 0 && !include.includes(project.name));
+
+      if (excludeProject) continue;
+
       if (projects.has(project.name)) {
         throw new CliError(
           `There are multiple projects with the same name [${project.name}]`,
```

(I also renamed to `toArray` just because I felt the intention was clearer)

kjbekkelund replied:
With this change we can also move more of the tests over to the `getProjects` tests.

kjbekkelund replied:
Oh, an improved `toArray` would of course be:

```ts
function toArray<T>(value?: T | T[]) {
  if (value == null) {
    return [];
  }

  if (Array.isArray(value)) {
    return value;
  }

  return [value];
}```

azasypkin replied:
> What do you think about moving this logic into getProjects?

Yeah, it looks much better! Initially I was thinking about use cases where we may need full/unfiltered list, but couldn't find any, so we should be good :)

azasypkin replied:
> With this change we can also move more of the tests over to the getProjects tests.

Yeah, but still have to keep a bunch of tests in `run.tests` to test string/string[] behavior for exclude/include filters :/

kjbekkelund commented on packages/kbn-pm/src/cli.ts:
```diff
@@ -23,6 +23,8 @@ function help() {
 
+       -e, --exclude        Exclude specified projects.
+       -i, --include        Include only specified projects. If left unspecified, it defaults to including all projects.
        --skip-kibana        Do not include the root Kibana project when running command.
```
We can do it in a follow-up, but feels like we can get rid of `--skip-kibana` now 🎉

azasypkin replied:
Yep, I'll prepare a follow-up with updated `README.md`.

elasticmachine commented:
## :broken_heart: Build Failed
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1728/)

elasticmachine commented:
## :green_heart: Build Succeeded
* [continuous-integration/kibana-ci/pull-request](https://kibana-ci.elastic.co/job/elastic-kibana-pull-request/1733/)

azasypkin commented:
6.x/6.3: 4e3bde555d274b6aec375a65115f1ae680ebc70a
